File: reverse_linked_list/reverse_linked_list.py
"""
Given the head of a singly linked list, reverse it in-place.
"""

# ---------------------------------------------------------
class Node:
    def __init__(self, val, next_=None):
        self.val = val
        self.next_ = next_

def build_linked_list(vals):
    head = Node(vals[0])
    curr_ = head.next_
    prev_ = head
    for val in vals[1:]:
        curr_ = Node(val)
        prev_.next_ = curr_
        prev_ = curr_
        curr_ = curr_.next_
    return head

def print_linked_list(head):
    print("[", end="")
    while head is not None:
        print(head.val, end="")
        head = head.next_
        if head is not None:
            print(", ", end="")
    print("]")

def reverse_linked_list(head):
    prev_ = None
    curr_ = head
    next_ = None
    while curr_ is not None:
        next_ = curr_.next_    # move next
        curr_.next_ = prev_    # reverse curr
        prev_ = curr_          # move prev
        curr_ = next_          # move curr
    head = prev_
    return head

# ---------------------------------------------------------
vals = [1,2,3,4]
head = build_linked_list(vals)
print_linked_list(head)
head = reverse_linked_list(head)
print_linked_list(head)
head = reverse_linked_list(head)
print_linked_list(head)
print()

vals = [1]
head = build_linked_list(vals)
print_linked_list(head)
head = reverse_linked_list(head)
print_linked_list(head)
print()

File: reverse_linked_list/reverse_linked_list.cpp
// g++ -std=c++17 reverse_linked_list.cpp
/*
Given the head of a singly linked list, reverse it in-place.
*/

#include <iostream>    // std::cout, std::endl
#include <vector>      // std::vector

// ------------------------------------------------------------------
struct Node {
    // Forward declarations can only be defined as pointers
    int val;
    Node* next_;
};

Node* newNode(int val) {
    Node* node = new Node;
    node->val = val;
    node->next_ = NULL;
    return node;
}

Node* build_linked_list(std::vector<int> vals) {
    // Build in reverse; insert to the front
    Node* head  = NULL;
    Node* curr_ = NULL;
    for (auto& val : vals) {
        curr_ = newNode(val);
        curr_->next_ = head;
        head = curr_;
    }
    return head;
}

void print_linked_list(Node* head) {
    std::cout << "[";
    while (head != NULL) {
        std::cout << head->val;
        head = head->next_;
        if (head != NULL)
            std::cout << ", ";
    }
    std::cout << "]" << std::endl;
}

Node* reverse_linked_list(Node* head) {
    Node* prev_ = NULL;
    Node* curr_ = head;
    Node* next_ = NULL;
    while (curr_ != NULL) {
        next_ = curr_->next_;    // move next
        curr_->next_ = prev_;    // reverse curr
        prev_ = curr_;           // move prev
        curr_ = next_;           // move curr
    }
    head = prev_;
    return head;
}

// ------------------------------------------------------------------
int main() {
    std::vector<int> vals;
    Node* head;

    vals = {4, 3, 2, 1};
    head = build_linked_list(vals);
    print_linked_list(head);
    head = reverse_linked_list(head);
    print_linked_list(head);
    head = reverse_linked_list(head);
    print_linked_list(head);
    std::cout << std::endl;

    vals = {1};
    head = build_linked_list(vals);
    print_linked_list(head);
    head = reverse_linked_list(head);
    print_linked_list(head);
    std::cout << std::endl;
}
